File: tools/validate-data.js
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { extractEmbeddedData, validateEmbeddedAnimeShape } from './lib/embedded-data.js';
import { buildTrailerUrls } from '../js/security/trailer-url-policy.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const DEFAULT_DATA_PATH = path.join(__dirname, '..', 'data', 'anime.full.json');
const DEFAULT_EMBEDDED_PATH = path.join(__dirname, '..', 'js', 'data.js');
const DEFAULT_INDEX_PATH = path.join(__dirname, '..', 'index.html');
const DEFAULT_BASELINE_PATH = path.join(__dirname, 'validation-baseline.json');
const NON_TOLERATED_BASELINE_ERRORS = new Set(['duplicateIds']);

const readJson = (filePath) => JSON.parse(fs.readFileSync(filePath, 'utf8'));

const parseEmbeddedDataScript = (embeddedPath) => {
  const source = fs.readFileSync(embeddedPath, 'utf8');
  return extractEmbeddedData(source);
};

const normalizeId = (anime) => {
  const meta = anime.metadata || {};
  return meta.id || anime.id || '';
};

const normalizeTitle = (anime) => {
  const meta = anime.metadata || {};
  return meta.title || anime.title || '';
};

const normalizeScore = (anime) => {
  const meta = anime.metadata || {};
  if (Object.prototype.hasOwnProperty.call(meta, 'score')) return meta.score;
  return anime.score;
};

const normalizeAniListId = (anime) => {
  const meta = anime.metadata || {};
  if (Object.prototype.hasOwnProperty.call(meta, 'anilistId')) return meta.anilistId;
  return anime.anilistId;
};

const normalizeTrailer = (anime) => {
  const meta = anime.metadata || {};
  return meta.trailer || anime.trailer || null;
};

const summarize = (title, groups) => {
  const lines = [];
  Object.entries(groups).forEach(([key, values]) => {
    if (!values.length) return;
    lines.push(`  ${key}: ${values.length}`);
  });
  if (!lines.length) {
    lines.push('  none');
  }
  return [`${title}:`, ...lines].join('\n');
};

const normalizeLabel = (value) => String(value || '').replace(/\\/g, '/');

const validateList = (animeList, label) => {
  const errors = {
    missingId: [],
    missingTitle: [],
    missingCover: [],
    missingScore: [],
    missingTrailer: [],
    invalidTrailer: [],
    invalidEpisodeScore: [],
    missingEpisodeScore: [],
    missingEpisodeNumber: [],
    duplicateIds: []
  };

  const warnings = {
    missingYear: [],
    missingSeason: [],
    missingStudio: [],
    missingSource: [],
    missingAnilistId: [],
    missingEpisodes: []
  };

  const idMap = new Map();

  animeList.forEach((anime, index) => {
    const meta = anime.metadata || {};
    const id = normalizeId(anime);
    const title = normalizeTitle(anime);
    const cover = meta.cover || anime.cover;
    const year = meta.year || anime.year;
    const season = meta.season || anime.season;
    const studio = meta.studio || anime.studio;
    const source = meta.source || anime.source;
    const score = normalizeScore(anime);
    const anilistId = normalizeAniListId(anime);
    const trailer = normalizeTrailer(anime);
    const episodes = anime.episodes || [];

    if (!id) {
      errors.missingId.push(index + 1);
    } else if (idMap.has(id)) {
      errors.duplicateIds.push(id);
    } else {
      idMap.set(id, true);
    }

    if (!title) errors.missingTitle.push(id || index + 1);
    if (!cover) errors.missingCover.push(id || index + 1);
    if (!Number.isFinite(Number(score))) errors.missingScore.push(id || index + 1);

    if (!year) warnings.missingYear.push(id || index + 1);
    if (!season) warnings.missingSeason.push(id || index + 1);
    if (!studio) warnings.missingStudio.push(id || index + 1);
    if (!source) warnings.missingSource.push(id || index + 1);
    if (!anilistId) warnings.missingAnilistId.push(id || index + 1);

    if (!Array.isArray(episodes) || episodes.length === 0) {
      warnings.missingEpisodes.push(id || index + 1);
    } else {
      episodes.forEach((ep) => {
        if (!ep || typeof ep !== 'object') return;
        if (ep.episode === undefined || ep.episode === null) {
          errors.missingEpisodeNumber.push(id || index + 1);
        }
        if (ep.score === undefined || ep.score === null) {
          errors.missingEpisodeScore.push(id || index + 1);
        } else if (!Number.isFinite(Number(ep.score)) || ep.score < 1 || ep.score > 5) {
          errors.invalidEpisodeScore.push(id || index + 1);
        }
      });
    }

    if (!trailer) {
      errors.missingTrailer.push(id || index + 1);
    } else {
      const { url, embedUrl } = buildTrailerUrls(trailer);
      if (!url && !embedUrl) {
        errors.missingTrailer.push(id || index + 1);
      } else if (!url || !embedUrl) {
        errors.invalidTrailer.push(id || index + 1);
      }
    }
  });

  const hasErrors = Object.values(errors).some((values) => values.length > 0);
  console.log(`Validation results (${label})`);
  console.log(summarize('Errors', errors));
  console.log(summarize('Warnings', warnings));
  console.log('');
  return { hasErrors, label, errors, warnings };
};

const validateIndexReferences = (indexPath) => {
  if (!fs.existsSync(indexPath)) {
    return { hasErrors: false, errors: [], warnings: [`Missing index file: ${indexPath}`] };
  }
  const html = fs.readFileSync(indexPath, 'utf8');
  const errors = [];
  const warnings = [];

  if (/const\s+ANIME_DATA\s*=/.test(html)) {
    errors.push('index.html still contains inline ANIME_DATA payload');
  }
  if (!/src=["']\/js\/main\.js["']/.test(html)) {
    warnings.push('index.html does not reference /js/main.js');
  }

  return { hasErrors: errors.length > 0, errors, warnings };
};

const parseArgs = (args) => {
  const values = {};
  const flags = new Set();

  for (let index = 0; index < args.length; index += 1) {
    const arg = args[index];
    if (!arg.startsWith('--')) continue;
    const [key, inlineValue] = arg.replace(/^--/, '').split('=');
    if (inlineValue !== undefined) {
      values[key] = inlineValue;
      continue;
    }
    const next = args[index + 1];
    if (next && !next.startsWith('--')) {
      values[key] = next;
      index += 1;
    } else {
      flags.add(key);
    }
  }

  return { values, flags };
};

const runValidation = ({
  dataPath = DEFAULT_DATA_PATH,
  embeddedPath = DEFAULT_EMBEDDED_PATH,
  indexPath = DEFAULT_INDEX_PATH,
  baselinePath = DEFAULT_BASELINE_PATH,
  enforceBaseline = false,
  skipEmbedded = false,
  skipIndexCheck = false
} = {}) => {
  const results = [];
  const data = readJson(dataPath);
  results.push(validateList(data.anime || [], normalizeLabel(path.relative(process.cwd(), dataPath))));

  if (!skipEmbedded) {
    try {
      const embedded = parseEmbeddedDataScript(embeddedPath);
      const shape = validateEmbeddedAnimeShape(embedded, { sampleSize: 50 });
      if (!shape.valid) {
        console.log(`Validation results (${normalizeLabel(path.relative(process.cwd(), embeddedPath))})`);
        console.log('Errors:');
        console.log(`  invalidEmbeddedShape: ${shape.errors.length}`);
        console.log('Warnings:');
        console.log('  none\n');
        results.push({
          hasErrors: true,
          label: normalizeLabel(path.relative(process.cwd(), embeddedPath)),
          errors: { invalidEmbeddedShape: shape.errors.map((_, index) => index) },
          warnings: {}
        });
      } else {
        results.push(validateList(embedded.anime || [], normalizeLabel(path.relative(process.cwd(), embeddedPath))));
      }
    } catch (error) {
      console.log(`Validation results (${normalizeLabel(path.relative(process.cwd(), embeddedPath))})`);
      console.log('Errors:');
      console.log('  invalidEmbeddedData: 1');
      console.log('Warnings:');
      console.log(`  detail: ${error.message}\n`);
      results.push({
        hasErrors: true,
        label: normalizeLabel(path.relative(process.cwd(), embeddedPath)),
        errors: { invalidEmbeddedData: [1] },
        warnings: {}
      });
    }
  }

  if (!skipIndexCheck) {
    const index = validateIndexReferences(indexPath);
    console.log(`Index reference check (${path.relative(process.cwd(), indexPath)})`);
    if (index.errors.length) {
      console.log('Errors:');
      index.errors.forEach((error) => console.log(`  - ${error}`));
    } else {
      console.log('Errors:\n  none');
    }
    if (index.warnings.length) {
      console.log('Warnings:');
      index.warnings.forEach((warning) => console.log(`  - ${warning}`));
    } else {
      console.log('Warnings:\n  none');
    }
    console.log('');
    results.push({
      hasErrors: index.hasErrors,
      label: normalizeLabel(path.relative(process.cwd(), indexPath)),
      errors: index.errors.reduce((acc, entry, indexPos) => ({ ...acc, [`indexError${indexPos + 1}`]: [entry] }), {}),
      warnings: index.warnings.reduce((acc, entry, indexPos) => ({ ...acc, [`indexWarning${indexPos + 1}`]: [entry] }), {})
    });
  }

  let baselineFailures = [];
  const baselineLabels = new Set();
  if (enforceBaseline) {
    if (!fs.existsSync(baselinePath)) {
      baselineFailures.push(`Missing baseline file: ${baselinePath}`);
    } else {
      const baseline = readJson(baselinePath);
      Object.keys(baseline || {}).forEach((label) => baselineLabels.add(String(label)));
      results.forEach((result) => {
        if (!result?.label) return;
        const baselineEntry = baseline?.[result.label];
        if (!baselineEntry) return;

        const compareGroup = (groupName, values = {}, expected = {}) => {
          const observedKeys = new Set([
            ...Object.keys(values || {}),
            ...Object.keys(expected || {})
          ]);
          observedKeys.forEach((key) => {
            const observed = Array.isArray(values[key]) ? values[key].length : 0;
            const limit = Number.isFinite(expected[key]) ? expected[key] : 0;
            if (observed > limit) {
              baselineFailures.push(
                `${result.label} ${groupName}.${key} exceeded baseline: ${observed} > ${limit}`
              );
            }
          });
        };

        compareGroup('errors', result.errors, baselineEntry.errors || {});
        compareGroup('warnings', result.warnings, baselineEntry.warnings || {});

        NON_TOLERATED_BASELINE_ERRORS.forEach((key) => {
          const allowed = Number(baselineEntry?.errors?.[key] ?? 0);
          const observed = Array.isArray(result?.errors?.[key]) ? result.errors[key].length : 0;
          if (allowed > 0) {
            baselineFailures.push(`${result.label} errors.${key} baseline allowance must be 0`);
          }
          if (observed > 0) {
            baselineFailures.push(`${result.label} errors.${key} must be 0 (observed ${observed})`);
          }
        });
      });
    }
  }

  if (baselineFailures.length) {
    console.log('Baseline regression check failed:');
    baselineFailures.forEach((entry) => console.log(`  - ${entry}`));
    console.log('');
  }

  const rawErrors = results.some((result) => {
    if (!result?.hasErrors) return false;
    if (!enforceBaseline) return true;
    if (!result.label) return true;
    return !baselineLabels.has(result.label);
  });

  const hasErrors = rawErrors || baselineFailures.length > 0;
  return { hasErrors, baselineFailures };
};

const main = () => {
  const { values, flags } = parseArgs(process.argv.slice(2));
  const result = runValidation({
    dataPath: values.data || DEFAULT_DATA_PATH,
    embeddedPath: values.embedded || DEFAULT_EMBEDDED_PATH,
    indexPath: values.index || DEFAULT_INDEX_PATH,
    baselinePath: values.baseline || DEFAULT_BASELINE_PATH,
    enforceBaseline: flags.has('enforce-baseline'),
    skipEmbedded: flags.has('skip-embedded'),
    skipIndexCheck: flags.has('skip-index-check')
  });
  if (result.hasErrors) {
    process.exitCode = 1;
  }
};

const isDirectRun = process.argv[1] && pathToFileURL(process.argv[1]).href === import.meta.url;
if (isDirectRun) {
  main();
}

export {
  parseEmbeddedDataScript,
  runValidation,
  validateIndexReferences,
  validateList
};

File: tools/validate_data.py
#!/usr/bin/env python3
"""Validate Rekonime catalog and embedded data."""

from __future__ import annotations

import argparse
import json
import os
from pathlib import Path
import re
import sys
from typing import Any
from urllib.parse import urlparse, urlunparse, parse_qsl, urlencode
from typing import Any, Dict, List, Optional, Set


ROOT = Path(__file__).resolve().parents[1]
DEFAULT_DATA_PATH = ROOT / "data" / "anime.full.json"
DEFAULT_EMBEDDED_PATH = ROOT / "js" / "data.js"
DEFAULT_INDEX_PATH = ROOT / "index.html"
DEFAULT_BASELINE_PATH = Path(__file__).resolve().parent / "validation-baseline.json"
NON_TOLERATED_BASELINE_ERRORS = {"duplicateIds"}
TRAILER_URL_HOSTS = {"youtube.com", "www.youtube.com", "m.youtube.com", "youtu.be"}
TRAILER_EMBED_HOSTS = {"youtube.com", "www.youtube.com", "youtube-nocookie.com", "www.youtube-nocookie.com"}


def read_json(path: Path) -> Any:
    return json.loads(path.read_text(encoding="utf-8"))


def normalize_label(value: str) -> str:
    return value.replace("\\", "/")


def rel_label(path: Path) -> str:
    return normalize_label(os.path.relpath(path, Path.cwd()))


def extract_embedded_data(source: str, identifier: str = "ANIME_DATA") -> Any:
    text = source.strip()
    if not text:
        raise ValueError("Embedded data script is empty")
    match = re.search(rf"(?:const|let|var)\s+{identifier}\s*=\s*([\s\S]+?);\s*$", text)
    if not match:
        raise ValueError(f"Unable to locate {identifier} payload")
    try:
        return json.loads(match.group(1))
    except json.JSONDecodeError as error:
        raise ValueError(f"Failed to parse {identifier} JSON payload: {error}") from error


def parse_embedded_data_script(path: Path) -> Any:
    return extract_embedded_data(path.read_text(encoding="utf-8"))


def is_plain_object(value: Any) -> bool:
    return isinstance(value, dict)


def validate_embedded_anime_shape(payload: Any, sample_size: int = 50) -> Dict[str, Any]:
    errors: List[str] = []
    anime_list = payload.get("anime") if isinstance(payload, dict) else None
    if not isinstance(anime_list, list):
        return {"valid": False, "errors": ["payload.anime must be an array"]}
    if not anime_list:
        return {"valid": True, "errors": []}
    limit = min(len(anime_list), max(1, sample_size))
    for index in range(limit):
        anime = anime_list[index]
        if not is_plain_object(anime):
            errors.append(f"anime[{index}] must be an object")
            continue
        if not isinstance(anime.get("genres"), list):
            errors.append(f"anime[{index}].genres must be an array")
        if not isinstance(anime.get("themes"), list):
            errors.append(f"anime[{index}].themes must be an array")
        if not isinstance(anime.get("episodes"), list):
            errors.append(f"anime[{index}].episodes must be an array")
        if anime.get("trailer") is not None and not is_plain_object(anime.get("trailer")):
            errors.append(f"anime[{index}].trailer must be an object or null")
        if not is_plain_object(anime.get("stats")):
            errors.append(f"anime[{index}].stats must be an object")
    return {"valid": len(errors) == 0, "errors": errors}


def normalize_id(anime: Dict[str, Any]) -> str:
    meta = anime.get("metadata") or {}
    return meta.get("id") or anime.get("id") or ""


def normalize_title(anime: Dict[str, Any]) -> str:
    meta = anime.get("metadata") or {}
    return meta.get("title") or anime.get("title") or ""


def normalize_score(anime: Dict[str, Any]) -> Any:
    meta = anime.get("metadata") or {}
    return meta["score"] if "score" in meta else anime.get("score")


def normalize_anilist_id(anime: Dict[str, Any]) -> Any:
    meta = anime.get("metadata") or {}
    return meta["anilistId"] if "anilistId" in meta else anime.get("anilistId")


def normalize_trailer(anime: Dict[str, Any]) -> Any:
    meta = anime.get("metadata") or {}
    return meta.get("trailer") or anime.get("trailer")


def sanitize_url(raw_url: Any, allowed_hosts: Set[str], remove_autoplay: bool = False) -> str:
    if not raw_url:
        return ""
    parsed = urlparse(str(raw_url))
    if parsed.scheme != "https":
        return ""
    if parsed.hostname not in allowed_hosts:
        return ""
    if remove_autoplay:
        query = [(key, value) for key, value in parse_qsl(parsed.query, keep_blank_values=True) if key != "autoplay"]
        parsed = parsed._replace(query=urlencode(query))
    return urlunparse(parsed)


def build_trailer_urls(trailer: Any) -> Dict[str, str]:
    if not isinstance(trailer, dict):
        return {"url": "", "embedUrl": ""}
    trailer_id = trailer.get("id")
    url = trailer.get("url") or ""
    embed_url = trailer.get("embedUrl") or trailer.get("embed_url") or ""
    if not url and trailer_id:
        url = f"https://www.youtube.com/watch?v={trailer_id}"
    if not embed_url and trailer_id:
        embed_url = f"https://www.youtube.com/embed/{trailer_id}"
    return {
        "url": sanitize_url(url, TRAILER_URL_HOSTS),
        "embedUrl": sanitize_url(embed_url, TRAILER_EMBED_HOSTS, remove_autoplay=True),
    }


def summarize(title: str, groups: Dict[str, List[Any]]) -> str:
    lines = [title + ":"]
    populated = [(key, values) for key, values in groups.items() if values]
    if not populated:
        lines.append("  none")
    else:
        lines.extend(f"  {key}: {len(values)}" for key, values in populated)
    return "\n".join(lines)


def validate_list(anime_list: List[Any], label: str) -> Dict[str, Any]:
    errors = {
        "missingId": [],
        "missingTitle": [],
        "missingCover": [],
        "missingScore": [],
        "missingTrailer": [],
        "invalidTrailer": [],
        "invalidEpisodeScore": [],
        "missingEpisodeScore": [],
        "missingEpisodeNumber": [],
        "duplicateIds": [],
    }
    warnings = {
        "missingYear": [],
        "missingSeason": [],
        "missingStudio": [],
        "missingSource": [],
        "missingAnilistId": [],
        "missingEpisodes": [],
    }
    id_map: Dict[str, bool] = {}

    for index, anime in enumerate(anime_list):
        item = anime if isinstance(anime, dict) else {}
        meta = item.get("metadata") or {}
        anime_id = normalize_id(item)
        title = normalize_title(item)
        cover = meta.get("cover") or item.get("cover")
        year = meta.get("year") or item.get("year")
        season = meta.get("season") or item.get("season")
        studio = meta.get("studio") or item.get("studio")
        source = meta.get("source") or item.get("source")
        score = normalize_score(item)
        anilist_id = normalize_anilist_id(item)
        trailer = normalize_trailer(item)
        episodes = item.get("episodes") or []
        fallback = anime_id or index + 1

        if not anime_id:
            errors["missingId"].append(index + 1)
        elif anime_id in id_map:
            errors["duplicateIds"].append(anime_id)
        else:
            id_map[anime_id] = True

        if not title:
            errors["missingTitle"].append(fallback)
        if not cover:
            errors["missingCover"].append(fallback)
        try:
            numeric_score = float(score)
        except (TypeError, ValueError):
            numeric_score = float("nan")
        if numeric_score != numeric_score:
            errors["missingScore"].append(fallback)

        if not year:
            warnings["missingYear"].append(fallback)
        if not season:
            warnings["missingSeason"].append(fallback)
        if not studio:
            warnings["missingStudio"].append(fallback)
        if not source:
            warnings["missingSource"].append(fallback)
        if not anilist_id:
            warnings["missingAnilistId"].append(fallback)

        if not isinstance(episodes, list) or not episodes:
            warnings["missingEpisodes"].append(fallback)
        else:
            for episode in episodes:
                if not isinstance(episode, dict):
                    continue
                if episode.get("episode") is None:
                    errors["missingEpisodeNumber"].append(fallback)
                if episode.get("score") is None:
                    errors["missingEpisodeScore"].append(fallback)
                else:
                    try:
                        episode_score = float(episode.get("score"))
                    except (TypeError, ValueError):
                        episode_score = float("nan")
                    if episode_score != episode_score or episode_score < 1 or episode_score > 5:
                        errors["invalidEpisodeScore"].append(fallback)

        if not trailer:
            errors["missingTrailer"].append(fallback)
        else:
            urls = build_trailer_urls(trailer)
            if not urls["url"] and not urls["embedUrl"]:
                errors["missingTrailer"].append(fallback)
            elif not urls["url"] or not urls["embedUrl"]:
                errors["invalidTrailer"].append(fallback)

    has_errors = any(len(values) > 0 for values in errors.values())
    print(f"Validation results ({label})")
    print(summarize("Errors", errors))
    print(summarize("Warnings", warnings))
    print("")
    return {"hasErrors": has_errors, "label": label, "errors": errors, "warnings": warnings}


def validate_index_references(index_path: Path) -> Dict[str, Any]:
    if not index_path.exists():
        return {"hasErrors": False, "errors": [], "warnings": [f"Missing index file: {index_path}"]}
    html = index_path.read_text(encoding="utf-8")
    errors: List[str] = []
    warnings: List[str] = []
    if re.search(r"const\s+ANIME_DATA\s*=", html):
        errors.append("index.html still contains inline ANIME_DATA payload")
    if not re.search(r"src=[\"']/js/main\.ts[\"']", html):
        warnings.append("index.html does not reference /js/main.ts")
    return {"hasErrors": bool(errors), "errors": errors, "warnings": warnings}


def run_validation(
    *,
    data_path: Path = DEFAULT_DATA_PATH,
    embedded_path: Path = DEFAULT_EMBEDDED_PATH,
    index_path: Path = DEFAULT_INDEX_PATH,
    baseline_path: Path = DEFAULT_BASELINE_PATH,
    enforce_baseline: bool = False,
    skip_embedded: bool = False,
    skip_index_check: bool = False,
) -> Dict[str, Any]:
    results: List[Dict[str, Any]] = []
    data = read_json(data_path)
    results.append(validate_list(data.get("anime") or [], rel_label(data_path)))

    if not skip_embedded:
        try:
            embedded = parse_embedded_data_script(embedded_path)
            shape = validate_embedded_anime_shape(embedded)
            if not shape["valid"]:
                print(f"Validation results ({rel_label(embedded_path)})")
                print("Errors:")
                print(f"  invalidEmbeddedShape: {len(shape['errors'])}")
                print("Warnings:")
                print("  none\n")
                results.append({
                    "hasErrors": True,
                    "label": rel_label(embedded_path),
                    "errors": {"invalidEmbeddedShape": list(range(len(shape["errors"])))},
                    "warnings": {},
                })
            else:
                results.append(validate_list(embedded.get("anime") or [], rel_label(embedded_path)))
        except Exception as error:
            print(f"Validation results ({rel_label(embedded_path)})")
            print("Errors:")
            print("  invalidEmbeddedData: 1")
            print("Warnings:")
            print(f"  detail: {error}\n")
            results.append({
                "hasErrors": True,
                "label": rel_label(embedded_path),
                "errors": {"invalidEmbeddedData": [1]},
                "warnings": {},
            })

    if not skip_index_check:
        index = validate_index_references(index_path)
        print(f"Index reference check ({rel_label(index_path)})")
        if index["errors"]:
            print("Errors:")
            for error in index["errors"]:
                print(f"  - {error}")
        else:
            print("Errors:\n  none")
        if index["warnings"]:
            print("Warnings:")
            for warning in index["warnings"]:
                print(f"  - {warning}")
        else:
            print("Warnings:\n  none")
        print("")
        results.append({
            "hasErrors": index["hasErrors"],
            "label": rel_label(index_path),
            "errors": {f"indexError{index_pos + 1}": [entry] for index_pos, entry in enumerate(index["errors"])},
            "warnings": {f"indexWarning{index_pos + 1}": [entry] for index_pos, entry in enumerate(index["warnings"])},
        })

    baseline_failures: List[str] = []
    baseline_labels: Set[str] = set()
    if enforce_baseline:
        if not baseline_path.exists():
            baseline_failures.append(f"Missing baseline file: {baseline_path}")
        else:
            baseline = read_json(baseline_path)
            baseline_labels = {str(label) for label in (baseline or {}).keys()}
            for result in results:
                label = result.get("label")
                baseline_entry = (baseline or {}).get(label)
                if not label or not baseline_entry:
                    continue
                for group_name in ["errors", "warnings"]:
                    values = result.get(group_name) or {}
                    expected = baseline_entry.get(group_name) or {}
                    for key in set(values.keys()) | set(expected.keys()):
                        observed = len(values.get(key) or []) if isinstance(values.get(key), list) else 0
                        limit = expected.get(key) if isinstance(expected.get(key), (int, float)) else 0
                        if observed > limit:
                            baseline_failures.append(f"{label} {group_name}.{key} exceeded baseline: {observed} > {limit}")
                for key in NON_TOLERATED_BASELINE_ERRORS:
                    allowed = int((baseline_entry.get("errors") or {}).get(key) or 0)
                    observed = len((result.get("errors") or {}).get(key) or [])
                    if allowed > 0:
                        baseline_failures.append(f"{label} errors.{key} baseline allowance must be 0")
                    if observed > 0:
                        baseline_failures.append(f"{label} errors.{key} must be 0 (observed {observed})")

    if baseline_failures:
        print("Baseline regression check failed:")
        for entry in baseline_failures:
            print(f"  - {entry}")
        print("")

    raw_errors = any(
        result.get("hasErrors")
        and (not enforce_baseline or not result.get("label") or result.get("label") not in baseline_labels)
        for result in results
    )
    return {"hasErrors": raw_errors or bool(baseline_failures), "baselineFailures": baseline_failures}


def main(argv: Optional[List[str]] = None) -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--data", default=str(DEFAULT_DATA_PATH))
    parser.add_argument("--embedded", default=str(DEFAULT_EMBEDDED_PATH))
    parser.add_argument("--index", default=str(DEFAULT_INDEX_PATH))
    parser.add_argument("--baseline", default=str(DEFAULT_BASELINE_PATH))
    parser.add_argument("--enforce-baseline", action="store_true")
    parser.add_argument("--skip-embedded", action="store_true")
    parser.add_argument("--skip-index-check", action="store_true")
    args = parser.parse_args(argv)

    result = run_validation(
        data_path=Path(args.data),
        embedded_path=Path(args.embedded),
        index_path=Path(args.index),
        baseline_path=Path(args.baseline),
        enforce_baseline=args.enforce_baseline,
        skip_embedded=args.skip_embedded,
        skip_index_check=args.skip_index_check,
    )
    return 1 if result["hasErrors"] else 0


if __name__ == "__main__":
    sys.exit(main())
